pkg/gce-pd-csi-driver/cache.go

```
@@ -547,3 +548,92 @@ func fetchChunkSizeKiB(cacheSize string) (string, error) {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		klog.V(2).ErrorS(err, "errored while creating watcher")
+	}
+	klog.V(2).Infof("Watcher started for directory %v", dirToWatch)
+	defer watcher.Close()
+
+	// out of the box fsnotify can watch a single file, or a single directory
+	if err := watcher.Add(dirToWatch); err != nil {
+		klog.V(2).ErrorS(err, "errored while adding watcher directory")
+	}
+	errorCh := make(chan error, 1)
+	// Handle the error received from the watcher goroutine
+	go watchDiskDetaches(watcher, nodeName, errorCh)
+
+	select {
+	case err := <-errorCh:
+		klog.Errorf("watcher encountered an error: %v", err)
+	}
+}
+
+func watchDiskDetaches(watcher *fsnotify.Watcher, nodeName string, errorCh chan error) error {
+	for {
+		select {
+		// watch for errors
+		case err := <-watcher.Errors:
+			errorCh <- fmt.Errorf("disk update event errored: %v", err)
+		// watch for events
+		case event := <-watcher.Events:
+			// In case of an event i.e. creation or deletion of any new PV, we update the VG metadata.
+			// This might include some non-LVM changes, no harm in updating metadata multiple times.
+			reduceVolumeGroup(getVolumeGroupName(nodeName), true)
+			klog.V(2).Infof("disk attach/detach event %#v\n", event)
+		}
+	}
+}
+
+func validateRaidedLSSDinVG(vgName string, lssdPath string) error {
+	args := []string{
+		"--noheadings",
+		"-o",
+		"pv_name",
+		"--select",
+		"vg_name=" + vgName,
+	}
+	info, err := common.RunCommand("" /* pipedCmd */, nil /* pipedCmdArg */, "pvs", args...)
+	if err != nil {
+		return fmt.Errorf("errored while checking physical volume details %v: %s", err, info)
+		// On error info contains the error message which we cannot use for further steps
+	}
+
+	if !strings.Contains(string(info), lssdPath) {
+		info, err := common.RunCommand("" /* pipedCmd */, nil /* pipedCmdArg */, "vgextend", []string{vgName, lssdPath}...)
+		if err != nil {
+			klog.Errorf("errored while extending VGs %v: %s", err, info)
```

We should return an error here. Also, can we separate the 2 functions? ValidateRaidedLSSD should only check if LSSD is present if not we can call some other function like `addRaidedSsdToVg`.

I am ok to refactor the code later but we should return error for RAID failures.

---

Done.